generator/deploy_openshift_pod.py

```diff
@@ -20,206 +20,361 @@
-                if resp.status.phase == "Succeeded":
-                    logger.info("All Containers in the Pod have terminated in success.")
-                    return True
-                # Wait for a second before another POD check
-                time.sleep(2)
-                count += 1
-                # If POD does not start during 10 seconds, then it probaby faild.
-                if count > 10:
-                    logger.error(
-                        "Deploying POD FAILED."
-                        "Either it does not start or it does not finished yet"
-                        "Status is: %r" % resp.status
+                    logger.info(
+                        "The pod has failed execution: you should "
+                        "inspect logs or check `oc describe`"
                     )
-                    return False
-        else:
-            logger.error(
-                f"POD with name {self.pod_name!r} "
-                f"already exists in namespace {self.project_name!r}"
-            )
-            return True
+                    break
+                if resp.status.phase == "Succeeded":
+                    logger.info("All Containers in the pod have finished successfully.")
+                    for m_dir in self.mapped_dirs:
+                        self.copy_path_to_pod(m_dir)
+                    break
+                # TODO: can we use watch instead?
+                time.sleep(1)
 
-    def deploy_image(self):
-        logger.info("Deploying image '%r' into a new POD.", self.upstream_name)
-        if "KUBERNETES_SERVICE_HOST" in os.environ:
-            self.create_pod_manifest()
-            self.setup_kubernetes()
-            result = self.deploy_pod()
-            self.delete_pod()
-            logger.info(f"Pod {self.pod_name!r} was deleted")
-            if not result:
-                logger.error("Running POD FAILED. Check generator logs for reason.")
-                return False
-            logger.info(
-                "Running POD was successful. "
-                f"Check {self.volume_dir} directory for results."
-            )
-            return True
-        else:
-            logger.warning("Generator IS NOT RUNNING in OpenShift.")
-            return False
+    def get_logs(self) -> str:
+        """ provide logs from the pod """
+        return self.api.read_namespaced_pod_log(
+            name=self.pod_name, namespace=self.k8s_namespace_name
+        )
+
+    def run(self, command: Optional[List] = None):
+        """
+        deploy a pod; if a command is set, we wait for it to finish,
+        if the command is not set, sleep process runs in the pod and you can use exec
+        to run commands inside
+
+        :param command: command to run in the pod, if None, run sleep
+        """
+        self.deploy_pod(command=command)
+        logger.info("Sandbox pod is deployed.")
+        return self.get_logs()
+
+    def exec(self, command: List[str]) -> str:
```

Comment:
later on, this should be a part of some `generator/api.py` file.

Reply:
ideally this class should be the api